contrib/man/mympd-config.1
.\" Manpage for mympd-config.
.\" Contact <mail@jcgames.de> to correct errors or typos.
.TH man 1 "11 Sep 2026" "26.1.0" "mympd-config man page"

.SH NAME
mympd-config \- mympd configuration tool

.SH SYNOPSIS
mympd-config [options...] [SETTING VALUE ...]

.SH DESCRIPTION
mympd-config is a small commandline tool to configure basic myMPD settings.

.SH OPTIONS
.TP
\fB-a </var/cache/mympd>\fR
Set's the path to the myMPD cache directory
.TP
\fB-b </usr/bin/mympd>\fR
Set's the path to the myMPD executable.
.TP
\fB-c </var/lib/mympd/config>\fR
Set's the path to the myMPD config directory.
.TP
\fB-s <0|1>\fR
0 = myMPD runs with normal linux user account
.br
1 = myMPD is started with systemd private user
.TP
\fB-a </var/lib/mympd>\fR
Set's the path to the myMPD working directory
.TP
\fBSETTING VALUE\fR
Valid settings are the environment variables respected by "mympd -c".

Reference: https://jcorporation.github.io/myMPD/020-configuration/configuration-files

.SH AUTHOR
Juergen Mang (mail@jcgames.de)

.SH BUGS
If you find a bug, please report it at https://github.com/jcorporation/myMPD/issues

.SH SEE ALSO
https://github.com/jcorporation/myMPD


CHANGELOG.md
# myMPD Changelog

https://github.com/jcorporation/myMPD/

***

## myMPD v26.1.0 (not yet released)

This is a small bug fix release.

### Changelog

- Upd: Embedded libmdpclient to 2.28 (master branch)
- Upd: Move playlist length info to top
- Upd: Optimize home icon layout
- Upd: Improve toolbar layout
- Fix: Show header for empty playlists

***

## myMPD v26.0.0 (2026-08-28)

This release removes the long-press event handling completely. It seems it can not be handled correctly in JavaScript for all mobile platforms and produces more headache than improving the user experience.

### Changelog

- Feat: Add home icon action badge
- Feat: Improve playlist shuffle performance
- Feat: Improve query parsing and add tests
- Feat: Serve locally installed documentation
- Upd: Improve url-encoding and -decoding
- Upd: Mongoose 7.23
- Upd: Improve websocket handling
- Upd: Embedded libmdpclient to 2.28 (master branch)
- Fix: Remove long-press events #1556
- Fix: Implement minimal signal save signal handler
- Fix: Ignore SIGPIPE
- Fix: Remove unnecessary dependency in mympd-script for libmpdclient #1559
- Fix: Path traversal checks
- Fix: Do not truncate 64bit integers

***

## myMPD v25.3.0 (2026-07-12)

This is a maintenance release.

### Changelog

- Upd: Translations
- Upd: Performance improvements for file reading functions
- Fix: Set libmpdclient min. version to 2.26.0
- Fix: Do not reset scrollposition on queue refresh
- Fix: Path traversal checks
- Fix: Do not leak fd on create_tmp_file failure

***

## myMPD v25.2.2 (2026-06-23)

This is a small bug fix release.

### Changelog

- Fix: Initialize placeholder images with default values
- Fix: libmpdclient bug

***

## myMPD v25.2.1 (2026-06-21)

This is a small bug fix release.

### Changelog

- Upd: Mongoose to current master
- Fix: OpenSSL 4.0 compatibility
- Fix: Invalid compile definitions

***

## myMPD v25.2.0 (2026-06-21)

This is a small feature release.

### Changelog

- Feat: Rework linked list implementation, it is now a double linked list
- Feat: Improve utf8 validation and handling
- Feat: Add WebradioDB update buttons to maintenance modal
- Upd: Test framework
- Upd: Embedded libmpdclient to current master
- Fix: Handle unexpected MPD disconnects
- Fix: Remove timers before removing partition
- Fix: Remove triggers before removing partition
- Fix: Harden MPD connection handling

***

## myMPD v25.1.1 (2026-06-03)

This is a small bug fix release.

### Changelog

- Upd: Add all mpd idle events to myMPD event list
- Upd: Translations
- Upd: Improve `mympd-config` script

***

## myMPD v25.1.0 (2026-05-31)

This is a small feature release.

### Changelog

- Feat: Add .jxl support for images #1540
- Feat: Show AudioFormat of streams in queue view #1547
- Feat: Add support for MPD_IDLE_NEIGHBOR and MPD_IDLE_MOUNT events
- Upd: Embedded libmpdclient to current master
- Upd: Mongoose to current master for TLS chain support
- Upd: Map locale zh-Hant to zh-Hans
- Fix: Build album cache also if database mtime has not changed for MPD mounts support #1545
- Fix: Keyboard navigation shortcut for "Show Playlists" #1548

***

## myMPD v25.0.2 (2026-05-03)

This is a small bug fix release.

### Changelog

- Upd: Mongoose 7.21 #1532
- Upd: Long-press-event to 2.5.2 #1538
- Upd: Trim whitespace for search strings #1535
- Fix: Restore scroll positions #1534

***

## myMPD v25.0.1 (2026-03-29)

This is a small bug fix release.

### Changelog

- Upd: Translation #1527 #1529
- Fix: Compile error with libmpdclient 2.24 #1528

***

## myMPD v25.0.0 (2026-03-22)

This is the first release that supports only MPD 0.23.5 and higher and Lua 5.4.x and higher.

### Changelog

- Feat: Use myGPIOd REST-API #1510
- Feat: Implement merge sort for linked lists
- Feat: Use a faster algorithm for shuffling linked lists
- Feat: Regularly save the myMPD state if myMPD is active
- Feat: Scripting - Add custom Lua function `mympd.firstTableValue`
- Upd: Bump requirement for MPD and Lua versions
- Upd: Search and utf8 handling improvements
- Upd: Add connection header to responses
- Upd: Improve HTTP session handling
- Fix: Check for minimum string length in json payload
- Fix: libutf8proc is an unused shared library in mympd-script #1520

***

## myMPD v24.0.3 (2026-02-10)

This is a small bug fix release.

### Changelog

- Upd: Split sds_extras compile unit
- Fix: test_utf8wrap still fails #1519
- Fix: Reset scrolling position on search

***

## myMPD v24.0.2 (2026-01-22)

This is a small bug fix release.

### Changelog

- Fix: Define NDEBUG for all release types but Debug #1515
- Fix: utf8 test failures #1514

***

## myMPD v24.0.1 (2026-01-06)

This is a small bug fix release.

### Changelog

- Fix: Handle invalid unicode strings #1511

***

## myMPD v24.0.0 (2025-12-30)

This release improves the integrated search by using string normalization and adding a fuzzy search option. Furthermore the mpd connection handling was improved.

The documentation site was migrated from Mkdocs to Sphinx, because of the deprecation of Material for Mkdocs.

### Changelog

- Feat: Fuzzy substring matching using the levenshtein distance
- Feat: String normalization for album, webradio, playlists and filesystem search
- Feat: Replace utf8 implementation with utf8proc library
- Feat: Add setting for default search operator
- Upd: Migrate documentation to Sphinx with Sphinx Book Theme #1495
- Upd: Move lyrics handling from mympd_api to webserver thread
- Upd: Performance improvements for mympd_api polling
- Upd: Stability improvements in MPD connection handling
- Upd: Limit length of smart playlists #1505
- Fix: Handling of HTTP connections #1503
- Fix: Endless scrolling in mobile view #1504

***

## myMPD v23.0.1 (2025-11-27)

This is a small bug fix release.

### Changelog

- Upd: Translations
- Upd: Mongoose 7.20
- Upd: Optimize build for openSUSE Build Service
- Fix: Segvault in album view if song title tag not exists
- Fix: Segvault in playlist view if song title tag not exists


cli_tools/mympd-config.c
/*
 SPDX-License-Identifier: GPL-2.0-or-later
 myMPD (c) 2018-2021 Juergen Mang <mail@jcgames.de>
 https://github.com/jcorporation/mympd
*/

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>
#include <inttypes.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "../dist/src/sds/sds.h"
#include "../src/sds_extras.h"

bool verbose = false;

struct t_config {
    //parsed values
    sds host;
    int port;
    sds pass;
    sds music_directory;
    sds playlist_directory;
    bool stickers;
    bool regex;
    //options
    sds mpd_conf;
    sds mpd_exe;
    sds mympd_conf;
    sds httphost;
    sds httpport;
    sds sslport;
    sds user;
    int loglevel;
    bool ssl;
    bool mixramp;
    bool publish;
    bool syscmds;
    bool readonly;
    bool scripting;
    bool chroot; /* Flawfinder: ignore */
};

int sdssplit_whitespace(sds line, sds *name, sds *value) {
    *name = sdsempty();
    *value = sdsempty();
    int tokens = 0;
    unsigned i = 0;
    const char *p = line;
    
    if (*p == '#') {
        return tokens;
    }
    //get name
    for (; i < sdslen(line); i++) {
        if (isspace(*p) != 0) {
            break;
        }
        else {
            *name = sdscatlen(*name, p, 1);
        }
        p++;
    }
    if (sdslen(*name) == 0) {
        return tokens;
    }
    else {
        tokens++;
    }
    //remove whitespace
    for (; i < sdslen(line); i++) {
        if (isspace(*p) == 0) {
            *value = sdscatlen(*value, p, 1);
        }
        p++;
    }
    //get value
    for (; i < sdslen(line); i++) {
        *value = sdscatlen(*value, p, 1);
        p++;
    }
    if (sdslen(*value) > 0) { tokens++; }
    sdstrim(*value, "\"");
    return tokens;
}

bool parse_mpd_conf(struct t_config *pconfig) {
    FILE *fp = fopen(pconfig->mpd_conf, "r");
    if (fp == NULL) {
        printf("Error parsing MPD configuration file: %s\n", pconfig->mpd_conf);
        return false;
    }
    printf("Parsing %s\n", pconfig->mpd_conf);
    char *line = NULL;
    size_t n = 0;
    sds sds_line = sdsempty();
    sds name;
    sds value;
    while (getline(&line, &n, fp) > 0) {
        sds_line = sdsreplace(sds_line, line);
        sdstrim(sds_line, " \n\r\t");
        if (sdslen(sds_line) > 0) {
            int tokens = sdssplit_whitespace(sds_line, &name, &value);
            if (tokens == 2) {
                if (strcasecmp(name, "music_directory") == 0) {
                    pconfig->music_directory = sdsreplace(pconfig->music_directory, value);
                    if (verbose) { printf("\tSetting music directory to %s\n", pconfig->music_directory); }
                }
                else if (strcasecmp(name, "playlist_directory") == 0) {
                    pconfig->playlist_directory = sdsreplace(pconfig->playlist_directory, value);
                    if (verbose) { printf("\tSetting playlist directory to %s\n", pconfig->playlist_directory); }
                }
                else if (strcasecmp(name, "sticker_file") == 0) {
                    if (sdslen(value) > 0) {
                        pconfig->stickers = true;
                        if (verbose) { printf("\tEnabling stickers\n"); }
                    }
                }
                else if (strcasecmp(name, "port") == 0) {
                    pconfig->port = strtoimax(value, NULL, 10);
                    if (verbose) { printf("\tSetting mpd port to %d\n", pconfig->port); }
                }
                else if (strcasecmp(name, "bind_to_address") == 0) {
                    if (sdslen(pconfig->host) == 0 || strncmp(value, "/", 1) == 0) {
                        //prefer socket connection
                        pconfig->host = sdsreplace(pconfig->host, value);
                        if (verbose) { printf("\tSetting mpd host to %s\n", pconfig->host); }
                    }
                }
                else if (strcasecmp(name, "password") == 0) {
                    sds *pwtokens;
                    int count;
                    pwtokens = sdssplitlen(value, strlen(value), "@", 1, &count);
                    if (count == 2) {
                        if (sdslen(pconfig->pass) == 0 || strstr(pwtokens[1], "admin") != NULL) {
                            //use prefered the entry with admin privileges or as fallback the first entry
                            pconfig->pass = sdsreplace(pconfig->pass, pwtokens[0]);
                            if (verbose) { printf("\tSetting mpd password\n"); }
                        }
                    }
                    sdsfreesplitres(pwtokens, count);
                }
            }
            sdsfree(name);
            sdsfree(value);
        }
    }
    if (line != NULL) {
        free(line);
    }
    fclose(fp);
    sdsfree(sds_line);
    return true;
}

bool check_ldd(struct t_config *pconfig) {
    printf("Getting mpd linked libraries %s\n", pconfig->mpd_exe);
    pconfig->regex = false;
    
    sds ldd = sdsnew("/usr/bin/ldd ");
    ldd = sdscatprintf(ldd, "%s", pconfig->mpd_exe);
    FILE *fp = popen(ldd, "r"); /* Flawfinder: ignore */
    sdsfree(ldd);
    if (fp == NULL) {
        return false;
    }
    char *line = NULL;
    size_t n = 0;
    while (getline(&line, &n, fp) > 0) {
        if (strstr(line, "libpcre") != NULL) {
            if (verbose) { printf("\tlibpcre found, enabling regex support\n"); }
            pconfig->regex = true;
            break;
        }
    }
    if (line != NULL) {
        free(line);
    }
    pclose(fp);
    return true;
}

bool parse_options(struct t_config *pconfig, int argc, char **argv) {
    int n, option_index = 0;
    
    static struct option long_options[] = {
        {"mpdconf",   required_argument, 0, 'c'},
        {"mpdexe",    required_argument, 0, 'e'},
        {"mympdconf", required_argument, 0, 'm'},
        {"httpport",  required_argument, 0, 'w'},
        {"sslport",   required_argument, 0, 's'},
        {"loglevel",  required_argument, 0, 'l'},
        {"user",      required_argument, 0, 'u'},
        {"mixramp",   no_argument,       0, 'r'},
        {"publish",   no_argument,       0, 'p'},
        {"syscmds",   no_argument,       0, 'y'},
        {"scripting", no_argument,       0, 't'},
        {"chroot",    no_argument,       0, 'o'},
        {"readonly",  no_argument,       0, 'n'},
        {"verbose",   no_argument,       0, 'v'},
        {"help",      no_argument,       0, 'h'},
        {0,           0,                 0,  0 }
    };

    while((n = getopt_long(argc, argv, "c:m:w:s:l:u:e:rpdhytov", long_options, &option_index)) != -1) { /* Flawfinder: ignore */
        switch (n) {
            case 'c':
                pconfig->mpd_conf = sdsreplace(pconfig->mpd_conf, optarg);
                break;
            case 'e':
                pconfig->mpd_exe = sdsreplace(pconfig->mpd_exe, optarg);
                break;
            case 'm':
                pconfig->mympd_conf = sdsreplace(pconfig->mympd_conf, optarg);
                break;
            case 'w':
                pconfig->httpport = sdsreplace(pconfig->httpport, optarg);
                break;
            case 'u':
                pconfig->user = sdsreplace(pconfig->user, optarg);
                break;
            case 'l':
                pconfig->loglevel = strtoimax(optarg, NULL, 10);
                break;
            case 's':
                pconfig->sslport = sdsreplace(pconfig->sslport, optarg);
                pconfig->ssl = strcmp(optarg, "0") == 0 ? false : true;
                break;
            case 'r':
                pconfig->mixramp = true;
                break;
            case 'p':
                pconfig->publish = true;
                break;
            case 'y':
                pconfig->syscmds = true;
                break;
            case 't':
                pconfig->scripting = true;
                break;
            case 'o':
                pconfig->chroot = true;
                break;
            case 'n':
                pconfig->readonly = true;
                break;
            case 'v':
                verbose = true;
                break;
            default:
                fprintf(stderr, "Usage: %s [OPTION]...\n"
                    "myMPD configuration utility, for details look at https://github.com/jcorporation/myMPD\n\n"
                    "-c, --mpdconf <path>     path to mpd.conf (default: /etc/mpd.conf)\n"
                    "-e, --mpdexe <path>      path to mpd executable (default: /usr/bin/mpd)\n"
                    "-m, --mympdconf <path>   path to mympd.conf (default: /etc/mympd.conf)\n"
                    "-w, --webport <port>     http port (default: 80)\n"
                    "-s, --sslport <port>     ssl port (default: 443, 0 to disable ssl)\n"
                    "-u, --user <username>    username to drop privileges to (default: mympd)\n"
                    "-l, --loglevel <number>  loglevel (default: 2 - info)\n"
                    "-r, --mixramp            enable mixramp settings in the gui (default: disabled)\n"
                    "-p, --publish            enable publishing feature (default: disabled)\n"
                    "-y, --syscmds            enable system commands (default: disabled)\n"
                    "-t, --scripting          enable scripting with lua (default: disabled)\n"
                    "-o, --chroot             enable chroot to /var/lib/mympd\n"
                    "-n, --readonly           enable readonly\n"
                    "-v, --verbose            enable verbose output\n"
                    "-h, --help               this help\n\n"
                    , argv[0]);
                return false;
        }
    }
    return true;
}

sds find_mpd_conf(void) {
    const char *filenames[] = { 
        "/etc/mpd.conf",
        "/usr/local/etc/mpd.conf",
        "/etc/opt/mpd/mpd.conf",
        "/etc/opt/mpd.conf",
        NULL
    };

    sds filename = sdsempty();
    for (const char **p = filenames; *p != NULL; p++) {
        filename = sdsreplace(filename, *p);
        if (access(filename, F_OK) == 0) { /* Flawfinder: ignore */
            printf("\tFound %s\n", filename);
            return filename;
        }
    }
    printf("\tNo mpd.conf found\n");
    filename = sdscrop(filename);
    return filename;
}

sds find_mpd_exe(void) {
    const char *filenames[] = { 
        "/usr/bin/mpd",
        "/usr/local/bin/mpd",
        "/opt/mpd/bin/mpd",
        NULL
    };

    sds filename = sdsempty();
    for (const char **p = filenames; *p != NULL; p++) {
        filename = sdsreplace(filename, *p);
        if (access(filename, F_OK) == 0) { /* Flawfinder: ignore */
            printf("\tFound %s\n", filename);
            return filename;
        }
    }
    printf("\tNo mpd binary found\n");
    filename = sdscrop(filename);
    return filename;
}

void set_defaults(struct t_config *pconfig) {
    printf("Searching for mpd\n");
    
    pconfig->host = sdsnew("/run/mpd/socket");
    pconfig->port = 6600;
    pconfig->pass = sdsempty();
    pconfig->music_directory = sdsempty();
    pconfig->playlist_directory = sdsempty();
    pconfig->stickers = false;
    pconfig->regex = true;
    pconfig->mpd_conf = find_mpd_conf();
    pconfig->mpd_exe = find_mpd_exe();
    pconfig->mympd_conf = sdsnew("/etc/mympd.conf");
    pconfig->httpport = sdsnew("80");
    pconfig->httphost = sdsnew("0.0.0.0");
    pconfig->sslport = sdsnew("443");
    pconfig->user = sdsnew("mympd");
    pconfig->loglevel = 5;
    pconfig->ssl = true;
    pconfig->mixramp = false;
    pconfig->publish = false;
    pconfig->syscmds = false;
    pconfig->scripting = false;
    pconfig->readonly = false;
    pconfig->chroot = false;
}

bool write_mympd_conf(struct t_config *pconfig) {
    printf("Writing %s\n", pconfig->mympd_conf);
    umask(0022);
    sds tmp_file = sdscatfmt(sdsempty(), "%s.XXXXXX", pconfig->mympd_conf);
    int fd = mkstemp(tmp_file);
    if (fd < 0) {
        fprintf(stderr, "Can't open %s for write\n", tmp_file);
        sdsfree(tmp_file);
        return false;
    }
    fchmod(fd, 0644);
    FILE *fp = fdopen(fd, "w");
    fprintf(fp, "# myMPD configuration file\n"
        "#\n"
        "# SPDX-License-Identifier: GPL-2.0-or-later\n"
        "# myMPD (c) 2018-2021 Juergen Mang <mail@jcgames.de>\n"
        "# https://github.com/jcorporation/mympd\n"
        "#\n\n"
        "[mpd]\n"
        "#Connection to mpd, unix socket or host/port, socket preferred\n"
        "#host = /run/mpd/socket\n"
        "#host = 127.0.0.1\n"
        "#port = 6600\n"
        "#pass = \n"
        "host = %s\n",
        pconfig->host);
    if (strncmp(pconfig->host, "/", 1) != 0) {
        fprintf(fp, "port = %d\n", pconfig->port);
    }
    if (sdslen(pconfig->pass) > 0) {
        fprintf(fp, "pass = %s\n", pconfig->pass);
    }
    fprintf(fp, "\n#absolut path of music_directory of mpd\n"
        "#none = no local music_directory\n"
        "#auto = get music_directory from mpd (only supported, if connected to mpd socket)\n");
    if (strncmp(pconfig->host, "/", 1) == 0) {
        fprintf(fp, "musicdirectory = auto\n");
    }
    else {
        fprintf(fp, "musicdirectory = %s\n", pconfig->music_directory);
    }
    fprintf(fp, "\n#absolut path of mpd playlist_directory\n"
        "playlistdirectory = %s\n\n", pconfig->playlist_directory);
    fprintf(fp, "#MPD compiled with regex support\n"
        "regex = %s\n", (pconfig->regex == true ? "true" : "false"));

    fprintf(fp, "\n\n[webserver]\n"
        "#Webserver options\n"
        "httphost = %s\n\n"
        "httpport = %s\n\n"
        "#Enable ssl\n"
        "#Certificates are generated under /var/lib/mympd/ssl/\n"
        "ssl = %s\n",
        pconfig->httphost,
        pconfig->httpport,
        (pconfig->ssl == true ? "true" : "false"));
    if (pconfig->ssl == true) {
        fprintf(fp, "sslport = %s\n", pconfig->sslport);
    }
    else {
        fprintf(fp, "#sslport = 443\n");
    }
    fprintf(fp, "\n#Publishes some mpd and myMPD directories\n"
        "publish = %s\n\n",
        (pconfig->publish == true ? "true" : "false"));

    fprintf(fp, "\n\n[mympd]\n"
        "Loglevel\n"
        "#0 = emerg\n"
        "#1 = alert\n"
        "#2 = crit\n"
        "#3 = error\n"
        "#4 = warn\n"
        "#5 = notice\n"
        "#6 = info\n"
        "#7 = debug\n"
        "loglevel = %d\n\n"
        "#myMPD user\n"
        "#group is the primary group of this user\n"
        "user = %s\n\n"
        "#Usage of stickers for play statistics\n"
        "stickers = %s\n\n"
        "#Mixrampdb settings in gui\n"
        "mixramp = %s\n\n"
        "#Enable system commands defined in syscmds section\n"
        "syscmds = %s\n\n"
        "#Enable lua scripting\n"
        "scripting = %s\n\n"
        "#Chroot to /var/lib/mympd\n"
        "chroot = %s\n"
        "readonly = %s\n",
        pconfig->loglevel,
        pconfig->user,
        (pconfig->stickers == true ? "true" : "false"),
        (pconfig->mixramp == true ? "true" : "false"),
        (pconfig->syscmds == true ? "true" : "false"),
        (pconfig->scripting == true ? "true" : "false"),
        (pconfig->chroot == true ? "true" : "false"),
        (pconfig->readonly == true ? "true" : "false"));
    
    fprintf(fp, "\n\n[syscmds]\n"
        "Shutdown = sudo /sbin/halt\n"
        "#To use this command add following lines to /etc/sudoers (without #)\n"
        "#Cmnd_Alias MYMPD_CMDS = /sbin/halt\n"
        "#mympd ALL=NOPASSWD: MYMPD_CMDS\n");
    
    fclose(fp);
    sds conf_file = sdscatfmt(sdsempty(), "%s", pconfig->mympd_conf);
    int rc = rename(tmp_file, conf_file);
    if (rc == -1) {
        fprintf(stderr, "Renaming file from %s to %s failed\n", tmp_file, conf_file);
    }
    sdsfree(tmp_file);
    sdsfree(conf_file);
    return true;
}

int main(int argc, char **argv) {
    int rc = EXIT_SUCCESS;
    
    struct t_config mpd_config;
    set_defaults(&mpd_config);

    if (parse_options(&mpd_config, argc, argv) == false) {
        rc = EXIT_FAILURE;
        goto cleanup;
    }

    if (sdslen(mpd_config.mpd_conf) > 0) {
        parse_mpd_conf(&mpd_config);
    }
 
    if (sdslen(mpd_config.mpd_exe) > 0) {
        check_ldd(&mpd_config);
    }

    write_mympd_conf(&mpd_config);

    cleanup:
    sdsfree(mpd_config.host);
    sdsfree(mpd_config.pass);
    sdsfree(mpd_config.music_directory);
    sdsfree(mpd_config.playlist_directory);
    sdsfree(mpd_config.httpport);
    sdsfree(mpd_config.httphost);
    sdsfree(mpd_config.sslport);
    sdsfree(mpd_config.user);
    sdsfree(mpd_config.mpd_conf);
    sdsfree(mpd_config.mpd_exe);
    sdsfree(mpd_config.mympd_conf);
    return rc;
}
